compiler/GlobalVariableReplacement.cpp

```diff
@@ -0,0 +1,364 @@
+                    }
+                } else {
+                    key = argv->expression->toString();
+                }
+                std::cout << "Here: key = " << key << std::endl;
+                if (write_flag_map[curr_state].count(key) != 0 && write_flag_map[curr_state][key] == 1) {
+                    // Start replacement for read part
+                    if (argv->expression->to<IR::PathExpression>()) {
+                        IR::PathExpression *up_path = 
+                        new IR::PathExpression(new const IR::Path(IR::ID("new_"+key+std::to_string(actual_write_replace_time_map[curr_state][key]))));
+                        actual_write_replace_time_map[curr_state][key]++;
+                        arguments_vec->at(i) = new IR::Argument(up_path);
+                        modify_flag = 1;
+                        write_flag_map[curr_state][key] = 0;
+                    } else if (auto mem = argv->expression->to<IR::Cast>()) {
+                        // new Cast(const IR::Type* type, const IR::Expression* expr);
+                        IR::Cast *up_cast = 
+                        new IR::Cast(mem->destType,new const IR::PathExpression(IR::ID("new_"+key+std::to_string(actual_write_replace_time_map[curr_state][key]))));
+                        actual_write_replace_time_map[curr_state][key]++;
+                        arguments_vec->at(i) = new IR::Argument(up_cast);
+                        modify_flag = 1;
+                        write_flag_map[curr_state][key] = 0;
+                    }
+                }
+            }
+        }
+        if (modify_flag == 0) {
+            return methodcall;
+        }
+        return new IR::MethodCallStatement(new IR::MethodCallExpression(call->method, call->typeArguments, arguments_vec));
+    }
+    
+};
+
+// This is invoked on each parser separately
+class UpdateParserByReplacingGlobalVar : public PassManager {
+
+ public:
+    explicit UpdateParserByReplacingGlobalVar(ReferenceMap *refMap, TypeMap *typeMap,
+    map_var_to_pos_appear* read_mp, map_var_to_pos_appear* write_mp, map_replace_width_mp* replace_width_mp) {
+        passes.push_back(new ComputeDepVar(refMap, typeMap, read_mp, write_mp, replace_width_mp));
+    }
+};
+
+}
+
+    bool FindReadWriteVariable::preorder(const IR::P4Parser *parser) {
+        CollectInfo collectIn(typeMap, read_mp, write_mp, replace_width_mp);
+        collectIn.setCalledBy(this);
+        parser->apply(collectIn);
+        return true;
+    }
+
+    const IR::Node *DoGlobalVariableReplacement::postorder(IR::P4Parser *parser) {
+        // std::cout << "Enter DoGlobalVariableReplacement::postorder(IR::P4Parser *parser)" << std::endl;
+        // for (auto &v : parser->states) {
+        //     for (auto &comp : v->components)
+        //         std::cout << "comp = " << comp << std::endl;
+        // }
+        // std::cout << "Exit DoGlobalVariableReplacement::postorder(IR::P4Parser *parser)" << std::endl;
```

Comment:
Maybe remove these comments?

Reply:
Since I have a follow up merge for the next pass, can I delete them in the next pass? In case I still need these printing messages for debugging before adding the next pass.

Reply:
Sure, no big issue.